<div align="center">
    <h1> My Machine Learning Journey</h1>
Civil Engineer · Researcher Youtuber · Machine Learning Engineer (The Goal)
    <br>
    Sharing and documenting my progress and journey as I learn machine learning
     <br>
</div>

<br>

<div align="center">
    <a target="_blank" href="https://github.com/barbaraaboagye/My-MachineLearning-Journey"><img src="https://img.shields.io/github/last-commit/barbaraaboagye/My-MachineLearning-Journey"></a>&nbsp;
      <a target="_blank" href="https://www.youtube.com/channel/UCEYKFq7ZEg81GYxpzNqYZ4)"><img src="https://img.shields.io/youtube/channel/subscribers/UCEYKFq7ZEg81GYxpzNqYZ4Q?style=social"></a>&nbsp;
    <a target="_blank" href="https://fr.linkedin.com/in/barbara-aboagye-233ba8133"><img src="https://img.shields.io/badge/style--5eba00.svg?label=LinkedIn&logo=linkedin&style=social"></a>&nbsp;
    <a target="_blank" href="https://twitter.com/awesome_ama"><img src="https://img.shields.io/twitter/follow/awesome_ama?style=social"></a>
    <br>
</div>

<br>

# PyTorch Foundation - Day x of x

Hours completed : 1hr 30min <br>

Finally, I was able to instal the torch. I just had to change the kernel to pytorch and voila it worked. I was using the wrong kernel all along that's why it kept on crushing even after I had Pytorch installed. 
> It was still crashing when I used spyder and Jupyter notebook. This solution worked for VSCODE. I didn't try to see if the other two will work :)

# Resources used : 

- [Foundation- PyTorch](https://madewithml.com/courses/foundations/pytorch/)
<br>

Handwritten notes can be found [here][def]

[def]: https://github.com/barbaraaboagye/My-MachineLearning-Journey/blob/9ea0e98265d73af4a028c43862b268230b1b37fb/Handwritten%20notes/230331_PytorchFoundations.pdf

### Set up 

In [39]:
import numpy as np
import torch

In [40]:
SEED = 1234

In [41]:
# Set seed for reproducibility

np.random.seed(seed=SEED)
torch.manual_seed(SEED)

### Basics 

We'll first cover some basics with PyTorch such as creating tensors and converting from common data structures (lists, arrays, etc.) to tensors.

In [42]:
# Creating a random tensor

x = torch.randn(2,3) # normal distribution (rand(2,3) -> uniform distribution)
print (f"Type: {x.type()}")
print (f"Size: {x.shape}")
print (f"Values: \n{x}")

Type: torch.FloatTensor
Size: torch.Size([2, 3])
Values: 
tensor([[ 0.0461,  0.4024, -1.0115],
        [ 0.2167, -0.6123,  0.5036]])


This code creates a tensor `x` of size `(2,3)` filled with random values from a normal distribution with mean 0 and standard deviation 1 using PyTorch.

The tensor `x` has 2 rows and 3 columns

In [43]:
#Zeros and Ones tensor

x = torch.zeros(2, 3)
print(x)
x = torch.ones(2,3)
print(x)

tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])


In [44]:
# List to Tensor

x = [1,2,3], [4,5,6 ] # normal list
print (x)
print(np.shape(x))
print(f"Type: {np.ndim(x)}")
y = torch.Tensor([[1,2,3], [4,5,6 ]])
print (f"Size: {y.shape}")
print(f"x:\n {y}")
z = torch.Tensor(x)
print (f"Size: {z.shape}")
print(f"x:\n {z}")

([1, 2, 3], [4, 5, 6])
(2, 3)
Type: 2
Size: torch.Size([2, 3])
x:
 tensor([[1., 2., 3.],
        [4., 5., 6.]])
Size: torch.Size([2, 3])
x:
 tensor([[1., 2., 3.],
        [4., 5., 6.]])


In [45]:
## NumPy array to Tensor

x = torch.Tensor(np.random.rand(2,3))
print (f"Size: {x.shape}")
print(f"x:\n {x}")

Size: torch.Size([2, 3])
x:
 tensor([[0.1915, 0.6221, 0.4377],
        [0.7854, 0.7800, 0.2726]])


In [46]:
# Changing tensor type 
x = torch.Tensor(3,4)
print(f"Type: {x.type()}")
print(x)
x = x.long()
print(f"Type: {x.type()}")
print(x)

Type: torch.FloatTensor
tensor([[5.3240e+22, 3.2689e+21, 2.1707e-18, 7.2128e+22],
        [2.1715e-18, 4.1488e-08, 1.0186e-11, 3.0878e-09],
        [6.6767e-07, 4.0519e-11, 2.7600e-06, 2.9571e-18]])
Type: torch.LongTensor
tensor([[-9223372036854775808, -9223372036854775808,                    0,
         -9223372036854775808],
        [                   0,                    0,                    0,
                            0],
        [                   0,                    0,                    0,
                            0]])


### Long tensor 

In programming, we often work with numbers, like 1, 2, 3, etc. However, sometimes we need to work with really large numbers that are too big to be represented using the normal integer data type.

For example, imagine you have a dataset with billions of entries, and you need to index into it using unique identifiers that are assigned to each entry. In this case, you would need a data type that can represent really large integer values, and that's where LongTensor comes in.

In PyTorch, a LongTensor is a special type of data structure that can hold large integer values ranging from -9223372036854775808 to 9223372036854775807. You can think of it like a container that can store really big numbers.

You can create a LongTensor by either explicitly converting an existing tensor of a different data type using the .long() method or by directly initializing it using the torch.LongTensor() constructor. Once you have a LongTensor, you can use it for various mathematical operations, just like any other tensor in PyTorch.

#### Operations

In [47]:
# Addition

x = torch.randn(2,3)
y = torch.randn(2,3)
z = x + y
print(f"Size: {z.shape}")
print(f"Values: \n{z}")

Size: torch.Size([2, 3])
Values: 
tensor([[ 0.0761, -0.6775, -0.3988],
        [ 3.0633, -0.1589,  0.3514]])
